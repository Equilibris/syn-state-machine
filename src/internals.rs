mod buffer;

pub use buffer::*;

pub trait ParserCursor {
    type Error;
}

pub struct LocError<'a, Loc>(pub &'a str, pub Loc);

pub trait Spanned {
    type Loc;

    fn span(&self) -> Self::Loc;
}

pub trait Finalizer<Out, With> {
    fn finalize(self, value: With) -> std::ops::ControlFlow<Out, Out>;
}

pub struct BlackHoleFinalizer<T>(pub T);

impl<Out, With> Finalizer<Out, With> for BlackHoleFinalizer<Out> {
    fn finalize(self, _: With) -> std::ops::ControlFlow<Out, Out> {
        std::ops::ControlFlow::Break(self.0)
    }
}

pub trait Parse<C: ParserCursor, With>: Sized {
    type Finalizer: Finalizer<Self, With>;

    fn parse(input: &mut ParseBuffer<C>) -> Result<Self::Finalizer, C::Error>;
}

pub trait Peek<C> {
    fn peek(input: &C) -> Option<usize>;
}
pub trait FixedPeek {
    const SKIP: usize;
}
pub trait PeekError<C: ParserCursor> {
    fn error(input: &C) -> C::Error;
}
pub trait CombineError<Other> {
    fn combine(&mut self, other: Other);
}
