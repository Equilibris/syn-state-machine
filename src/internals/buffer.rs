use crate::internals::*;

#[derive(Clone, Copy)]
pub struct ParseBuffer<C> {
    pub cursor: C,
}

impl<C> From<C> for ParseBuffer<C> {
    fn from(cursor: C) -> Self {
        Self { cursor }
    }
}

impl<C: Iterator + ParserCursor> ParseBuffer<C> {
    pub fn peek<T: Peek<C>>(&mut self) -> bool {
        if let Some(x) = T::peek(&self.cursor) {
            let _ = self.cursor.advance_by(x);
            true
        } else {
            false
        }
    }
    pub fn errored_peek<T: Peek<C> + PeekError<C>>(&mut self) -> Result<(), C::Error> {
        if let Some(x) = T::peek(&self.cursor) {
            let _ = self.cursor.advance_by(x);
            Ok(())
        } else {
            Err(T::error(&self.cursor))
        }
    }
}
impl<C: ParserCursor> ParseBuffer<C> {
    pub fn parse<T: Parse<C, ()>>(&mut self) -> Result<T, C::Error> {
        Ok(match T::parse(self)?.finalize(()) {
            std::ops::ControlFlow::Continue(v) => v,
            std::ops::ControlFlow::Break(v) => v,
        })
    }
    pub fn call<T, E>(&mut self, function: impl Fn(&mut Self) -> Result<T, E>) -> Result<T, E> {
        function(self)
    }
}

impl<C: Spanned> Spanned for ParseBuffer<C> {
    type Loc = C::Loc;

    fn span(&self) -> C::Loc {
        self.cursor.span()
    }
}

impl<C> AsRef<C> for ParseBuffer<C> {
    fn as_ref(&self) -> &C {
        &self.cursor
    }
}
