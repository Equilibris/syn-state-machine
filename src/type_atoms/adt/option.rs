use crate::internals::*;

impl<T: Parse> Parse for Option<T> {
    fn parse<'a>(input: &ParseBuffer<'a>) -> Result<(Self, ParseBuffer<'a>)> {
        let mut input = input.clone();

        Ok((
            match input.parse() {
                Ok(a) => Some(a),
                Err(_) => None,
            },
            input,
        ))
    }
}
impl<T: Peek> Peek for Option<T> {
    fn peek<'a>(input: &ParseBuffer<'a>) -> Option<usize> {
        Some(T::peek(input).unwrap_or_default())
    }
}

#[cfg(test)]
mod tests {
    use crate::*;

    insta_match_test!(it_matches_only, Option<Ident> : <);
    insta_match_test!(it_returns_the_correct_length, Option<(Ident, Ident)> : hi <);
}
