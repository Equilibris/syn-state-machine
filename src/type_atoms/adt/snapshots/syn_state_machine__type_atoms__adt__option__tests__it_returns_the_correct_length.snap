---
source: src/type_atoms/adt/option.rs
expression: "(insta_match_test!(+ parse Option<(Ident, Ident)> : hi <),\n    insta_match_test!(* peek Option<(Ident, Ident)> : hi <))"
---
(
    Ok(
        None,
    ),
    Some(
        0,
    ),
)
